pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/AvroContainerTypeRecord.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.sink.pulsar;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.Lists;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.apache.pulsar.client.api.schema.Field;
+import org.apache.pulsar.client.api.schema.GenericRecord;
+import org.apache.pulsar.common.schema.SchemaType;
+
+/** Wraps Avro container types (i.e. Maps and Lists) in a generic record * */
+public class AvroContainerTypeRecord implements GenericRecord {
+  private final List<Field> fields;
+  private final Object nativeObject;
+
+  public AvroContainerTypeRecord(JsonNode nativeObject) {
+    if (nativeObject == null) throw new NullPointerException("nativeObject cannot be null!");
+    if (!nativeObject.isContainerNode())
+      throw new UnsupportedOperationException("nativeObject must be a container node!");
+
+    this.fields = populateFields(nativeObject);
+    this.nativeObject = nativeObject;
+  }
+
+  private AvroContainerTypeRecord(JsonNode nativeObject, List<Field> fields) {
+    this.fields = fields;
+    this.nativeObject = nativeObject;
+  }
+
+  @Override
+  public byte[] getSchemaVersion() {
+    throw new UnsupportedOperationException("GenericRecordWrapper does not support schema version");
+  }
+
+  @Override
+  public List<Field> getFields() {
+    return fields;
+  }
+
+  @Override
+  public Object getField(String fieldName) {
+    JsonNode fn = ((JsonNode) this.nativeObject).get(fieldName);
+    if (fn == null) {
+      return null;
+    } else if (fn.isContainerNode()) {
+      List<Field> fields = populateFields(fn);
+      return new AvroContainerTypeRecord(fn, fields);
+    } else {
+      return fn.isNull() ? null : fn.asText();
```

Comment:
how do we deal with numeric types ?
asText() will convert everything to a string

Reply:
AvroContainerTypeRecord is only valid for container types (Object or Array JsonNodeType). This is validated at the public constructor. fn.asText(); should never happen, I just ported it from JsonGenericRecord and serves as last resort.

Reply:
Then we have to throw a IllegalStateException

Reply:
I misspoke, the types here depend on whats in the map. You are right about numeric type, let me port the complete function from GenericJsonRecord.